### ISSUE TITLE
[iOS] Add ability to know reason why the map camera changed

### ISSUE THREAD

**Issue by nitaliano** [mapbox/mapbox-gl-native#10674]
Currently the Android SDK provides a callback for us to know the reason why the map has changed. https://www.mapbox.com/android-docs/api/map-sdk/5.2.1/com/mapbox/mapboxsdk/maps/MapboxMap.OnCameraMoveStartedListener.html#field.summary it would be awesome if we can add a feature similar to this in the iOS SDK. GL JS also passes along an `originalEvent` when the region changes so the developer will actually know why the map is changing.

/cc @lilykaiser

**Comment by 1ec5** [comment]
There are already a couple ways to distinguish programmatic camera changes from user-initiated camera changes:

* `-mapView:shouldChangeFromCamera:toCamera:` is only called for user-initiated camera changes, since it makes little sense to ask the developer “are you sure?” when programmatically changing the camera.
* That method doesn’t indicate which gesture recognized triggered the camera change, but you can distinguish gestures from each other by:
  * Finding the gesture recognizer you’re interested in among `MGLMapView.gestureRecognizers` and adding your own action to it.
  * Installing your own gesture recognizer that doesn’t prevent the built-in gesture recognizes from failing.
* Methods such as `-setCamera:withDuration:animationTimingFunction:completionHandler:` call a completion handler after the requested camera change or after any drift animation completes. Code within the completion block has access to the code that initiated the camera change.

**Comment by 1ec5** [comment]
#11151 implements an undocumented API that exposes a reason for each should/will/is/did delegate method. We left it undocumented because there are a number of open questions around how and whether this information should be exposed to a general developer audience. [As with any undocumented API](https://github.com/mapbox/mapbox-gl-native/wiki/Versions-and-tagging#mapbox-ios-sdk), it’s possible to use the API but the API is subject to change without notice and without a semver-major bump.

To get this feature to a point where we’d be comfortable making it public, it would have to be rewritten to separate cause (programmatic versus gesture, and which gesture) from effect (what kind of change took place in the camera). https://github.com/mapbox/mapbox-gl-native/pull/11151#issuecomment-365017014 outlines a clean redesign of this feature:

* `-mapView:shouldChangeFromCamera:toCamera:` becomes `-mapView:shouldChangeFromCamera:to:Camera:inResponseToGestureRecognizer:`.
* `-mapView:regionWillChangeAnimated:` becomes `-mapView:willChangeFromCamera:toCamera:animated:inResponseToGestureRecognizer:`
* `-mapViewRegionIsChanging:` becomes `-mapView:isChangingFromCamera:toCamera:inResponseToGestureRecognizer:`.
* `-mapView:regionDidChangeAnimated:` becomes `-mapView:didChangeFromCamera:toCamera:animated:inResponseToGestureRecognizer:`.

A more minimal alternative would be to separate the `reason` parameter into `reason` and `effect` parameters (or combining them into a `context` struct or dictionary).

**Comment by saschpe** [comment]
There is even more to it. Let's assume one wants to reverse geocode whenever the map centered on a new position. That would be the case:

1) Gesture by user
2) Programmatic position update (i.e. after some other user interface interaction)
3) Location component tracking update by the SDK

The only thing that reliably works today is 1) with `REASON_API_GESTURE`. `REASON_DEVELOPER_ANIMATION` and `REASON_API_ANIMATION` are essentially the same and number 3) is entirely missing.

For the outlined use-case, one would want to reverse geocode in case 1) and 2) but never 3) (which would lead to an endless loop of requests quickly eating up your budget ;-)). A way to implement this currently looks like this:

```kotlin
map.addOnCameraMoveStartedListener { reason ->
    if (reason == REASON_API_GESTURE) { cameraMoveHappened = true }
}
map.addOnCameraIdleListener {
    if (cameraMoveHappened) {
        reverseGeocodingSearch()
        cameraMoveHappened = false
    }
}

// And ...

private fun enableLocationComponent() {
    // ...
    mapboxMap?.locationComponent?.apply {
        cameraMoveHappened = true
        activateLocationComponent(locationComponentActivationOptions)
    }
}
```

Long story short, the reasons need an update (and should use an enum):

- **REASON_GESTURE**
- **REASON_API**
- **REASON_TRACKING**
